### PR TITLE
Switch the cursor to a busy cursor while redrawing.

### PR DESCRIPTION
**Pull request by anntzer**
To test, plot e.g. `plot(rand(100000))` and zoom in manually (the
redraw should take some time).  The cursor should switch to a "busy"
cursor (e.g. spinwheel).  The switch doesn't seem to happen under the
Tk and WX backends, probably due to some event loop intricacies I don't
understand.

### CODE REVIEW

**Comment by anntzer** [comment]
On an vaguely related note, I noticed the methods `Axes.{get,set}_cursor_props`, which seem to have no effect and have never been used (according to `git log -S`).  Would it be OK to just get rid of them or should we go through a deprecation cycle as usual?

**Comment by WeatherGod** [comment]
... or we make it actually do something. It is fully functional and
documented... it's state just doesn't seem to get accessed anywhere else,
though.

**Comment by tacaswell** [comment]
I am curious of the history when those methods were added and why.  There are a surprising number if bread crumb trails in the codebase.

**Comment by tacaswell** [comment]
Also, this is a _huge_ UX improvement this should be backported to 2.0

**Comment by anntzer** [comment]
Perhaps you're better than me at `git log -{G,S}` but at least I didn't find anything :-(  Just seems like a completely unused attribute.
Who are the Tk and WX specialists there?  The feature doesn't work with these backends (nothing's broken but the cursor doesn't change) so it'd be nice but not obligatory to fix that first.

**Comment by QuLogic** [comment]
Are the Cairo backends also handled here?

**Comment by anntzer** [comment]
Good catch.  A quick look seems to indicate that the Cairo backend does not override `draw`; where should the cursor change go?

**Comment by tacaswell** [comment]
Maybe on the `on_draw_event` in gtk3cairo and `expose_event` in gtk_cairo?

**Comment by anntzer** [comment]
Done.  The cursor switches to a spinwheel, but is not animated, for a reason I don't understand.

Side notes:
- The GTK3Cairo backend prints the warning "(-c:14995): Gtk-WARNING **: Allocating size to GtkToolItem 0x5564c90a5490 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?" when picking the zoom tool.
- It would probably be worth upping the minimum PyGTK version requirement to something like 2.12 (2009) to get rid of conditional branches that support old APIs.

**Comment by tacaswell** [comment]
@OceanWolf @fariza can you address the gtk questions?

File: lib/matplotlib/backends/backend_agg.py [hunk 1]
```diff
@@ -460,9 +460,14 @@ def draw(self):
         # acquire a lock on the shared font cache
         RendererAgg.lock.acquire()
 
+        toolbar = self.toolbar
         try:
+            if toolbar:
+                toolbar.set_cursor(cursors.WAIT)
```
Comment (fariza) [inline lib/matplotlib/backends/backend_agg.py]: This is going to fail with new tool manager, toolbar doesn't have a set_cursor method anymore.
I have to think in a clean way of doing it.

File: lib/matplotlib/backends/backend_agg.py [hunk 2]
```diff
@@ -460,9 +460,14 @@ def draw(self):
...
             self.figure.draw(self.renderer)
         finally:
+            if toolbar:
+                toolbar.set_cursor(cursors.POINTER)
```
Comment (fariza) [inline lib/matplotlib/backends/backend_agg.py]: why do you go back to `pointer`? you should go back to the cursor that was in use before

Comment (anntzer) [inline lib/matplotlib/backends/backend_agg.py]: Good catch, but AFAICT there's no `get_cursor` method available...
Actually from the point of view of Qt the best solution would probably be to use http://doc.qt.io/qt-5/qguiapplication.html#setOverrideCursor (because the entire GUI is blocked during the redraw (... actually there may be some argument for doing the drawing asynchronously I guess)) but I don't know whether other toolkits support something similar.

**Comment by anntzer** [comment]
Right now it looks like Qt4Agg Qt5Agg GtkCairo and Gtk3Cairo all work (I don't see the warning mentioned in https://github.com/matplotlib/matplotlib/pull/6603#issuecomment-227040771 anymore).  In theory Tk and WX "should" be supported too but don't actually work.  Could we push this forward with just support for these backends?

**Comment by QuLogic** [comment]
Ping @fariza @anntzer; what should we do about this? Seems useful...

**Comment by anntzer** [comment]
Rebased.  Now works on py3+qt4 qt5 tk gtk3 and py2 wx gtk (haven't tested other combos).  See the latest commit for the reason (we needed to trigger the event loop on some backends).

**Comment by tacaswell** [comment]
@fariza What owns the cursor in the new tool model?  It does make a certain amount of sense for the toolbar to do so (as it holds the other things that change the state of the cursor.  The other logical option is for the manager to own it.

Not sure I like putting this all the way down in `Agg`, seems like it should be fully owned by the GUI layer.

A decorator like:

```python
def wating_cursor(meth):
      @contextlib.contextmanager
      def spin_cursor(manager):
          manager and manager.set_cursor(cursor.WAIT)
          try:
              yield
         finally:
            manager and manager.set_cursor(cursor.POINTER)
   

     @functools.wraps(meth)
     def inner(self, *args, **kwargs):
        mager = getattr(self, 'manager', None)
        with spin_cursor(manager):
            return meth(self, *args, **kwarg)
```

would make this easy and safe to apply.

**Comment by fariza** [comment]
@anntzer is the problem of returning to previously selected cursor solved?

@tacaswell the cursors doesn't belong to the toolbar, right now there is a tool that sets the cursor when in axes `backend_tools.SetCursorBase._set_cursor_cbk`.

We can just trigger an event to inform the tool that we are waiting. But this seems a little bit overkill and maybe too much resources?

**Comment by anntzer** [comment]
Now it is :)

**Comment by anntzer** [comment]
@tacaswell I'm not convinced that adding a contextmanager around each and every call to draw() from an interactive backend is the solution (e.g. people sometimes call figure.canvas.draw() themselves)... or I guess interactive backends could derive from a common InteractiveCanvas class (earlier than FigureCanvasAgg in the MRO) which defines `draw()` as
```
def draw(...):
    <sets cursor>
    super().draw(...)
    <unsets cursor>
```
but then you're *actually* relying on multiple inheritance to do your job correctly (which I am actually totally fine with).

File: lib/matplotlib/backends/backend_gtk3.py [hunk 3]
```diff
@@ -538,7 +539,7 @@ def set_message(self, s):
 
     def set_cursor(self, cursor):
         self.canvas.get_property("window").set_cursor(cursord[cursor])
-        #self.canvas.set_cursor(cursord[cursor])
+        Gtk.main_iteration()
```
Comment (fariza) [inline lib/matplotlib/backends/backend_gtk3.py]: why are you adding this cal to `Gtk.main_iteration`?

**Comment by fariza** [comment]
For the toolmanager, I think it can be done using the `trigger` method of the `ToolSetCursor`, we can use that method to override the active cursor...  I will try to find some time this week to give it a try.

File: lib/matplotlib/backends/backend_gtk3.py (hunk 3 above)
Comment (anntzer) [inline lib/matplotlib/backends/backend_gtk3.py]: Because otherwise the cursor is actually not set on the GUI (see docstring changed in CanvasBase.set_cursor).  Basically, what happens is

Notification that a redraw is necessary.
Redraw starts and cursor is set.
<possibly long computation>
Return to GUI thread.

Note that the GUI event loop doesn't get to run between <cursor is set> and <possibly long computation>, so depending on the details of the toolkit, the cursor change may not actually appear to the user before the end of the computation.  Letting the GUI event loop run once allows it to change the cursor.  (And upon testing, I see that Qt does not need this, probably because its version of setCursor triggers(?) the event loop.)

**Comment by fariza** [comment]
@tacaswell what would be the problem moving `set_cursor` from the `toolbar` to the `firgurecanvas`? it seems a more logical place for it to be, and more useful.

**Comment by anntzer** [comment]
(fwiw I agree that it may make sense to allow ourselves to change the cursor even if there is no toolbar (ie in some embedding cases))

**Comment by tacaswell** [comment]
The canvas class makes more sense than the manager class so that direct embeddings can take advantage of it, but we will have to be careful to not mess with anything end-user may already be doing with the cursor.

**Comment by anntzer** [comment]
We are already messing up with the cursor if the embedding includes a
toolbar.
We could e.g. say "mpl may modify the cursor state.  If you do not wish
this to happen, override the canvas.set_cursor method to lambda cursor:
None".


On Jun 11, 2017 7:47 PM, "Thomas A Caswell" <notifications@github.com>
wrote:

The canvas class makes more sense than the manager class so that direct
embeddings can take advantage of it, but we will have to be careful to not
mess with anything end-user may already be doing with the cursor.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/matplotlib/matplotlib/pull/6603#issuecomment-307678533>,
or mute the thread

**Comment by tacaswell** [comment]
Fair enough.

**Comment by anntzer** [comment]
Rebased.

File: lib/matplotlib/backends/backend_agg.py (hunk 1 above)
Comment (WeatherGod) [inline lib/matplotlib/backends/backend_agg.py]: On a completely different note... you are doing this to the AGG backend, a non-interactive backend. How does this code interact with the combined backends like GtkAgg and such? Would this code get called twice?

**Comment by fariza** [comment]
So we are not moving the `set_cursor` to the canvas? I think this is the right time and place to do it

**Comment by tacaswell** [comment]
We can do that later, I think it is worth getting the busy-spinner in now.

**Comment by choldgraf** [comment]
is this ready to go? looks like features are there...time to merge this 1-year-old PR? :-)

File: lib/matplotlib/backends/backend_agg.py (hunk 1 above)
Comment (WeatherGod) [inline lib/matplotlib/backends/backend_agg.py]: I never got an answer about why this is being done for the AGG backend.

Comment (OceanWolf) [inline lib/matplotlib/backends/backend_agg.py]: Aye, but from what I gather set_cursor will move from the toolbar to the canvas.

I'm +1 on doing this now so that we get a clean PR, but @tacaswell disagrees.

**Comment by anntzer** [comment]
(I think this is good to go.)

File: lib/matplotlib/backends/backend_agg.py (hunk 1 above)
Comment (WeatherGod) [inline lib/matplotlib/backends/backend_agg.py]: @OceanWolf, was that a response to my question? I am not sure how your response applies, since the AGG backend has a canvas, too. Why is there any cursor-handling code for the agg backend?

Comment (OceanWolf) [inline lib/matplotlib/backends/backend_agg.py]: Yes, It applies only in the sense that the code here has already been slated for removal.  Only a partial answer to your question.

Comment (anntzer) [inline lib/matplotlib/backends/backend_agg.py]: Sorry, missed the comments.
The reason to put this here is the same reason why canvas.toolbar exists on all canvases (possibly None) rather than only on interactive ones: because there is no InteractiveCanvas class in the hierarchy :-)  (Note that this is orthogonal with the toolmanager question)

**Comment by anntzer** [comment]
rebased

**Comment by anntzer** [comment]
Allowing myself to mark this as release critical as it's been around for very long, it's been approved, and @tacaswell says it's "a *huge* UX improvement" :-)

**Comment by tacaswell** [comment]
I am happy this went in :+1:
